### ISSUE TITLE
High memory usage in http/rpc framework

### ISSUE THREAD

**Issue by harshavardhana** [minio/minio#6251]
** not using the template on purpose **

After using Minio from the latest master and doing some performance testing while reading 32MiB objects with 4096 workers, we see a significant increase in memory. Erasure coding is not using a lot of memory here instead it seems that memory is significantly higher due to gob.NewDecoder using bytes.Buffer. 

The relevant graph attached here... [mem.pdf](https://github.com/minio/minio/files/2264057/mem.pdf)

**Comment by harshavardhana** [comment]
[mem.pdf](https://github.com/minio/minio/files/2264234/mem.pdf) - made a slight change in cmd/rpc to see the implementation difference, still there is large amount of memory usage. 

```diff
git diff
diff --git a/cmd/rpc/server.go b/cmd/rpc/server.go
index 675c26352..66e45cafb 100644
--- a/cmd/rpc/server.go
+++ b/cmd/rpc/server.go
@@ -249,13 +249,9 @@ func (server *Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
                callResponse.Error = err.Error()
        }
 
-       data, err := gobEncode(callResponse)
-       if err != nil {
-               w.WriteHeader(http.StatusInternalServerError)
-               return
-       }
+       gob.NewEncoder(w).Encode(callResponse)
 
-       w.Write(data)
+       w.(http.Flusher).Flush()
 }
 
 // NewServer - returns new RPC server.
```

**Comment by harshavardhana** [comment]
The significant memory usage has resulted from two functions i.e 

```go
func gobEncode(e interface{}) ([]byte, error) {
        var buf bytes.Buffer

        if err := gob.NewEncoder(&buf).Encode(e); err != nil {
                return nil, err
        }

        return buf.Bytes(), nil
}
```

and RPC client call method

```go
        callRequest := CallRequest{
                Method:   serviceMethod,
                ArgBytes: data,
        }

        var buf bytes.Buffer
        if err = gob.NewEncoder(&buf).Encode(callRequest); err != nil {
                return err
        }
```

Notice the usage of bytes.Buffer without sync.Pool results in significant GC overhead.

**Comment by harshavardhana** [comment]
Should be fixed by 65de2d68c00b2f9cc265035b8176991557affa6d
